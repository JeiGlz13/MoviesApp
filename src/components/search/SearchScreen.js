import React, { useMemo } from 'react';
import queryString from 'query-string';

import { useLocation } from 'react-router-dom';
import { useForm } from '../hooks/useForm';
import { MovieCard } from '../movies/MovieCard';
import { getMoviesByName } from '../selectors/getMoviesByName';
import { Titles } from '../UI/Titles';

import searchMovieIcon from '../../assets/SearchMovie.webp';
import notFound from '../../assets/NotFound.png';

export const SearchScreen = ({history}) => {
    const location = useLocation();
    
    const {q = ''} = queryString.parse(location.search);
    const initialForm = {
        //name del input
        searchText: q
    }
    const [formValues, handleInputChange] = useForm(initialForm);
    
    const {searchText} = formValues;

    const moviesFiltered = useMemo(() => 
         getMoviesByName(q), [q]);
    
    const handleSearch = (e)=>{
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return (
        <div className='w-full px-10'>
            <Titles title='Buscar Pelicula' />

            <div className = "flex flex-row mt-3" >
                <div className = "w-full" >
                    <h4 className='text-sky-600 text-3xl font-extrabold mb-4' >Search Form</h4>

                    <div className='flex flex-row justify-center items-center mt-6' >
                        <form onSubmit = {handleSearch}
                         className="relative text-gray-700 w-96 ">
                            <input type="search"
                            name = "searchText"
                            value = {searchText}
                            onChange = {handleInputChange}
                            autoComplete = "off"
                            placeholder="Search"
                            className="bg-gray-100 w-full h-12 px-7 pr-12 rounded-full 
                            text-lg focus:outline-none cardShadow"/>
                        
                            <button type="submit" className="absolute right-0 top-0 mt-[0.65rem] mr-5">
                                <i className="fa-solid fa-magnifying-glass text-lg"></i>
                            </button>
                        </form>
                    </div>
                    
                </div>

                <div className = "w-full justify-center flex flex-col ml-14" >
                    <h4 className='text-sky-600 text-3xl font-extrabold mb-3' >Results</h4>

                    {   (q==='') 
                        && <div className = "flex flex-col justify-center items-center">
                                <img src={searchMovieIcon} alt='search ' className='w-72' />
                                <h1 className='font-extrabold text-3xl' >Search a movie</h1>
                            </div>
                    }
                    {   (q!=='' && moviesFiltered.length ===0) 
                        && <div className = "flex flex-col justify-center items-center">
                                <img src={notFound} alt='search ' className='w-72' />
                                <h1 className='font-extrabold text-3xl text-red-500' >Movie Not Found</h1>
                            </div>
                    }
                    <div className='flex flex-wrap' >
                        {
                            moviesFiltered.map(movie => (
                                <MovieCard
                                key={movie.id}
                                {...movie} />
                            ))
                        }
                    </div>
                </div>
            </div>
        </div>
    )
}
