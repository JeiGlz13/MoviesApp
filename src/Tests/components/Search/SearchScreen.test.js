import React from 'react';
import {mount} from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';
describe('Pruebas en <SearchScreen/>', () => {
    
    test('Debe mostrarse correctamente con valores por defecto', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path = "/search" component = {SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar pelicula');
    });

    test('Debe mostrar Casino y el input con el valor del querystring', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Casino']}>
                <Route path = "/search" component = {SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('Casino');
    });

    test('Debe mostrar un error si no se encuentra la pelicula', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Casino3']}>
                <Route path = "/search" component = {SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').exists()).toBe(true);
        expect(wrapper.find('.alert-danger').text().trim()).toBe('No se encontro Casino3');
    });
    
    test('Debe llamar el push del history', () => {
        const history = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Casino3']}>
                <Route path = "/search" component = {() => <SearchScreen history = {history} />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {target: {
            name: 'searchText',
            value: 'casino'
        }});

        wrapper.find('form').prop('onSubmit')({preventDefault(){}});
        expect(history.push).toHaveBeenCalledWith('?q=casino');
    })
    
    
})
