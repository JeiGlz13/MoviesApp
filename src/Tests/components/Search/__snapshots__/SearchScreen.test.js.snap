// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar pelicula"
          type="text"
          value=""
        />
      </form>
      <button
        className="btn m-1 btn-block btn-outline-primary"
        type="submit"
      >
        Buscar...
      </button>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <div
        className="alert alert-info"
      >
        Buscar pelicula
      </div>
    </div>
  </div>
</div>
`;
